File: src/main/kotlin/br/com/kurtis/payment_details/PaymentMessageCodec.java
package br.com.kurtis.payment_details;

import io.vertx.core.buffer.Buffer;
import io.vertx.core.eventbus.MessageCodec;
import io.vertx.core.json.Json;

public class PaymentMessageCodec implements MessageCodec<Payment, Payment> {

  @Override
  public void encodeToWire(final Buffer buffer, final Payment payment) {
    final String string = Json.encode(payment);
    final int length = string.getBytes().length;
    buffer.appendInt(length);
    buffer.appendString(string);
  }

  @Override
  public Payment decodeFromWire(final int position, final Buffer buffer) {
    final int length = buffer.getInt(position);
    final int start = position + 4;
    final int end = start + length;

    final String string = buffer.getString(start, end);
    return Json.decodeValue(string, Payment.class);
  }

  @Override
  public Payment transform(final Payment payment) {
    return payment;
  }

  @Override
  public String name() {
    return "payment";
  }

  @Override
  public byte systemCodecID() {
    return -1;
  }
}

File: src/main/kotlin/br/com/kurtis/payment_details/PaymentMessageCodec.kt
package br.com.kurtis.payment_details

import io.vertx.core.buffer.Buffer
import io.vertx.core.eventbus.MessageCodec
import io.vertx.core.json.Json

class PaymentMessageCodec : MessageCodec<Payment, Payment> {

  override fun encodeToWire(buffer: Buffer, payment: Payment) {
    val string = Json.encode(payment)
    val length = string.toByteArray().size
    buffer.appendInt(length)
    buffer.appendString(string)
  }

  override fun decodeFromWire(position: Int, buffer: Buffer): Payment {
    val length = buffer.getInt(position)
    val start = position + 4
    val end = start + length

    val string = buffer.getString(start, end)
    return Json.decodeValue(string, Payment::class.java)
  }

  override fun transform(payment: Payment): Payment {
    return payment
  }

  override fun name(): String {
    return "payment"
  }

  override fun systemCodecID(): Byte {
    return -1
  }
}
